# Activity: Perform feature engineering 

## **Introduction**


As you're learning, data professionals working on modeling projects use featuring engineering to help them determine which attributes in the data can best predict certain measures.

In this activity, you are working for a firm that provides insights to the National Basketball Association (NBA), a professional North American basketball league. You will help NBA managers and coaches identify which players are most likely to thrive in the high-pressure environment of professional basketball and help the team be successful over time.

To do this, you will analyze a subset of data that contains information about NBA players and their performance records. You will conduct feature engineering to determine which features will most effectively predict whether a player's NBA career will last at least five years. The insights gained then will be used in the next stage of the project: building the predictive model.


## **Step 1: Imports** 


Start by importing `pandas`.

In [1]:
# Import pandas
import pandas as pd



The dataset is a .csv file named `nba-players.csv`. It consists of performance records for a subset of NBA players. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [3]:
# Load the dataset into a variable named 'data'
data = pd.read_csv("nba-players.csv", index_col=0)

# Display the first few rows to ensure it's loaded correctly
data.head()

,name,gp,min,pts,fgm,fga,fg,3p_made,3pa,3p,...,fta,ft,oreb,dreb,reb,ast,stl,blk,tov,target_5yrs
0,Brandon Ingram,36,27.4,7.4,2.6,7.6,34.7,0.5,2.1,25.0,...,2.3,69.9,0.7,3.4,4.1,1.9,0.4,0.4,1.3,0
1,Andrew Harrison,35,26.9,7.2,2.0,6.7,29.6,0.7,2.8,23.5,...,3.4,76.5,0.5,2.0,2.4,3.7,1.1,0.5,1.6,0
2,JaKarr Sampson,74,15.3,5.2,2.0,4.7,42.2,0.4,1.7,24.4,...,1.3,67.0,0.5,1.7,2.2,1.0,0.5,0.3,1.0,0
3,Malik Sealy,58,11.6,5.7,2.3,5.5,42.6,0.1,0.5,22.6,...,1.3,68.9,1.0,0.9,1.9,0.8,0.6,0.1,1.0,1
4,Matt Geiger,48,11.5,4.5,1.6,3.0,52.4,0.0,0.1,0.0,...,1.9,67.4,1.0,1.5,2.5,0.3,0.3,0.4,0.8,1


<details><summary><h4><strong>Hint 1</strong></h4></summary>

The `read_csv()` function from `pandas` allows you to read in data from a csv file and load it into a DataFrame.
    
</details>

<details><summary><h4><strong>Hint 2</strong></h4></summary>

Call the `read_csv()`, pass in the name of the csv file as a string, followed by `index_col=0` to use the first column from the csv as the index in the DataFrame.
    
</details>

## **Step 2: Data exploration** 

Display the first 10 rows of the data to get a sense of what it entails.

In [4]:
# Display first 10 rows of the data
data.head(10)


,name,gp,min,pts,fgm,fga,fg,3p_made,3pa,3p,...,fta,ft,oreb,dreb,reb,ast,stl,blk,tov,target_5yrs
0,Brandon Ingram,36,27.4,7.4,2.6,7.6,34.7,0.5,2.1,25.0,...,2.3,69.9,0.7,3.4,4.1,1.9,0.4,0.4,1.3,0
1,Andrew Harrison,35,26.9,7.2,2.0,6.7,29.6,0.7,2.8,23.5,...,3.4,76.5,0.5,2.0,2.4,3.7,1.1,0.5,1.6,0
2,JaKarr Sampson,74,15.3,5.2,2.0,4.7,42.2,0.4,1.7,24.4,...,1.3,67.0,0.5,1.7,2.2,1.0,0.5,0.3,1.0,0
3,Malik Sealy,58,11.6,5.7,2.3,5.5,42.6,0.1,0.5,22.6,...,1.3,68.9,1.0,0.9,1.9,0.8,0.6,0.1,1.0,1
4,Matt Geiger,48,11.5,4.5,1.6,3.0,52.4,0.0,0.1,0.0,...,1.9,67.4,1.0,1.5,2.5,0.3,0.3,0.4,0.8,1
5,Tony Bennett,75,11.4,3.7,1.5,3.5,42.3,0.3,1.1,32.5,...,0.5,73.2,0.2,0.7,0.8,1.8,0.4,0.0,0.7,0
6,Don MacLean,62,10.9,6.6,2.5,5.8,43.5,0.0,0.1,50.0,...,1.8,81.1,0.5,1.4,2.0,0.6,0.2,0.1,0.7,1
7,Tracy Murray,48,10.3,5.7,2.3,5.4,41.5,0.4,1.5,30.0,...,0.8,87.5,0.8,0.9,1.7,0.2,0.2,0.1,0.7,1
8,Duane Cooper,65,9.9,2.4,1.0,2.4,39.2,0.1,0.5,23.3,...,0.5,71.4,0.2,0.6,0.8,2.3,0.3,0.0,1.1,0
9,Dave Johnson,42,8.5,3.7,1.4,3.5,38.3,0.1,0.3,21.4,...,1.4,67.8,0.4,0.7,1.1,0.3,0.2,0.0,0.7,0


<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

There is a function in the `pandas` library that can be called on a DataFrame to display the first n number of rows, where n is a number of your choice. 
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Call the `head()` function and pass in 10.
</details>

Display the number of rows and the number of columns to get a sense of how much data is available to you.

In [5]:
# Display number of rows and columns
data.shape


(1340, 21)

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

DataFrames in `pandas` have an attribute that can be called to get the number of rows and columns as a tuple.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

You can call the `shape` attribute.
</details>

**Question:** What do you observe about the number of rows and the number of columns in the data?

Once you execute this, you'll get the dimensions of the dataset (rows and columns). Based on that, you can describe the dataset's structure, such as how many observations (rows) and features (columns) it has.

Now, display all column names to get a sense of the kinds of metadata available about each player. Use the columns property in pandas.


In [6]:
# Display all column names
data.columns



Index(['name', 'gp', 'min', 'pts', 'fgm', 'fga', 'fg', '3p_made', '3pa', '3p',
       'ftm', 'fta', 'ft', 'oreb', 'dreb', 'reb', 'ast', 'stl', 'blk', 'tov',
       'target_5yrs'],
      dtype='object')

The following table provides a description of the data in each column. This metadata comes from the data source, which is listed in the references section of this lab.

<center>

|Column Name|Column Description|
|:---|:-------|
|`name`|Name of NBA player|
|`gp`|Number of games played|
|`min`|Number of minutes played per game|
|`pts`|Average number of points per game|
|`fgm`|Average number of field goals made per game|
|`fga`|Average number of field goal attempts per game|
|`fg`|Average percent of field goals made per game|
|`3p_made`|Average number of three-point field goals made per game|
|`3pa`|Average number of three-point field goal attempts per game|
|`3p`|Average percent of three-point field goals made per game|
|`ftm`|Average number of free throws made per game|
|`fta`|Average number of free throw attempts per game|
|`ft`|Average percent of free throws made per game|
|`oreb`|Average number of offensive rebounds per game|
|`dreb`|Average number of defensive rebounds per game|
|`reb`|Average number of rebounds per game|
|`ast`|Average number of assists per game|
|`stl`|Average number of steals per game|
|`blk`|Average number of blocks per game|
|`tov`|Average number of turnovers per game|
|`target_5yrs`|1 if career duration >= 5 yrs, 0 otherwise|

</center>

Next, display a summary of the data to get additional information about the DataFrame, including the types of data in the columns.

In [7]:
# Display a summary of the DataFrame using .info()
data.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1340 non-null   object 
 1   gp           1340 non-null   int64  
 2   min          1340 non-null   float64
 3   pts          1340 non-null   float64
 4   fgm          1340 non-null   float64
 5   fga          1340 non-null   float64
 6   fg           1340 non-null   float64
 7   3p_made      1340 non-null   float64
 8   3pa          1340 non-null   float64
 9   3p           1340 non-null   float64
 10  ftm          1340 non-null   float64
 11  fta          1340 non-null   float64
 12  ft           1340 non-null   float64
 13  oreb         1340 non-null   float64
 14  dreb         1340 non-null   float64
 15  reb          1340 non-null   float64
 16  ast          1340 non-null   float64
 17  stl          1340 non-null   float64
 18  blk          1340 non-null   float64
 19  tov   

**Question:** Based on the preceding tables, which columns are numerical and which columns are categorical?

Based on the `.info()` output:

- **Numerical columns**: These columns contain continuous or discrete numerical data.
  - `gp` (Games played)
  - `min` (Minutes played per game)
  - `pts` (Average number of points per game)
  - `fgm` (Average number of field goals made per game)
  - `fga` (Average number of field goal attempts per game)
  - `fg` (Average percentage of field goals made per game)
  - `3p_made` (Average number of three-point field goals made per game)
  - `3pa` (Average number of three-point field goal attempts per game)
  - `3p` (Average percentage of three-point field goals made per game)
  - `ftm` (Average number of free throws made per game)
  - `fta` (Average number of free throw attempts per game)
  - `ft` (Average percentage of free throws made per game)
  - `oreb` (Average number of offensive rebounds per game)
  - `dreb` (Average number of defensive rebounds per game)
  - `reb` (Average number of rebounds per game)
  - `ast` (Average number of assists per game)
  - `stl` (Average number of steals per game)
  - `blk` (Average number of blocks per game)
  - `tov` (Average number of turnovers per game)
  - `target_5yrs` (1 if career duration >= 5 years, 0 otherwise)

- **Categorical columns**: These columns contain categorical or qualitative data.
  - `name` (Name of NBA player)



### Check for missing values

Now, review the data to determine whether it contains any missing values. Begin by displaying the number of missing values in each column. After that, use isna() to check whether each value in the data is missing. Finally, use sum() to aggregate the number of missing values per column.


In [8]:
# Display the number of missing values in each column
missing_values = data.isna().sum()

# Check whether each value is missing
missing_check = data.isna()

# Aggregate the number of missing values per column
missing_values


name           0
gp             0
min            0
pts            0
fgm            0
fga            0
fg             0
3p_made        0
3pa            0
3p             0
ftm            0
fta            0
ft             0
oreb           0
dreb           0
reb            0
ast            0
stl            0
blk            0
tov            0
target_5yrs    0
dtype: int64

**Question:** What do you observe about the missing values in the columns? 

It seems that you would like to provide your own observations about the missing values in the dataset. Once you've run the code to check for missing values, you can write your observations based on the output.

Here's how you can interpret the result:

- If there are **no missing values** (i.e., all columns have a count of 0 missing values), you can observe that the dataset is complete and ready for analysis.
- If **some columns have missing values**, you can specify which columns they are and how many missing values are in each.

For example, a response could be:
- "There are no missing values in the dataset, meaning all columns are complete."
- "The `pts` and `fgm` columns have missing values, with 15 and 20 entries missing respectively."



**Question:** Why is it important to check for missing values?

It is important to check for missing values because:

Data Integrity: Missing values can affect the quality of the data, leading to inaccurate analysis or incorrect model predictions.

Data Preprocessing: Identifying missing values is crucial before applying any data transformations or feeding the data into a machine learning model.

Model Performance: If missing values are not handled appropriately, they could degrade the performance of the model or lead to biased results. Missing data can lead to unreliable conclusions if the data is not managed properly.

Data Cleaning: Identifying missing values allows us to choose the right strategy to handle them, such as imputing missing values, removing rows/columns, or using algorithms that can handle missing data.

## **Step 3: Statistical tests** 



Next, use a statistical technique to check the class balance in the data. To understand how balanced the dataset is in terms of class, display the percentage of values that belong to each class in the target column. In this context, class 1 indicates an NBA career duration of at least five years, while class 0 indicates an NBA career duration of less than five years.

In [9]:
# Display percentage of values for each class (1, 0) in the target column 'target_5yrs'
class_balance = data['target_5yrs'].value_counts(normalize=True) * 100

# Display the percentage for each class
class_balance


1    62.014925
0    37.985075
Name: target_5yrs, dtype: float64

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

In `pandas`, `value_counts(normalize=True)` can be used to calculate the frequency of each distinct value in a specific column of a DataFrame.  
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

After `value_counts(normalize=True)`, multipling by `100` converts the frequencies into percentages (%).
</details>

**Question:** What do you observe about the class balance in the target column?

It seems like you are looking to provide your observations based on the class balance. Once you run the code to check the class balance in the `target_5yrs` column, you will be able to write your response.

Here’s a possible response based on the output:

- "If the percentages of class 1 (NBA career duration of at least 5 years) and class 0 (NBA career duration of less than 5 years) are significantly different, the dataset might be imbalanced. For example, if class 0 has 80% of the data and class 1 has only 20%, it indicates a class imbalance."

Once you run the code, you can describe the balance between these classes and any imbalances that might affect the analysis or model building.

Feel free to share the output, and I can help you with the interpretation!

**Question:** Why is it important to check class balance?


It is important to check class balance because:

1. **Model Bias**: An imbalanced dataset can cause the model to become biased toward the majority class. If one class is overrepresented, the model might learn to predict that class most of the time, leading to poor performance for the minority class.

2. **Accuracy is Not Enough**: In imbalanced datasets, accuracy can be misleading. For instance, if 95% of the samples belong to one class, a model predicting only that class could achieve 95% accuracy, even if it completely fails to predict the minority class. In such cases, other metrics like precision, recall, and F1-score become more important.

3. **Prediction Quality**: A model trained on imbalanced data may fail to generalize well and could make inaccurate predictions for the underrepresented class. This is especially critical in scenarios where identifying the minority class is crucial, such as in fraud detection, medical diagnoses, or rare event prediction.

4. **Resampling Needs**: Checking class balance helps identify whether resampling techniques (like oversampling the minority class or undersampling the majority class) are necessary to improve model performance and prevent bias towards the majority class.

5. **Fairness**: Imbalanced classes can lead to fairness issues, especially when predictions are used for decision-making in real-world scenarios. The model may underperform for certain groups, potentially leading to inequitable outcomes.

Understanding and addressing class balance ensures the development of more accurate, reliable, and fair predictive models.

## **Step 4: Results and evaluation** 


Now, perform feature engineering, with the goal of identifying and creating features that will serve as useful predictors for the target variable, `target_5yrs`. 

### Feature selection

The following table contains descriptions of the data in each column:

<center>

|Column Name|Column Description|
|:---|:-------|
|`name`|Name of NBA player|
|`gp`|Number of games played|
|`min`|Number of minutes played|
|`pts`|Average number of points per game|
|`fgm`|Average number of field goals made per game|
|`fga`|Average number of field goal attempts per game|
|`fg`|Average percent of field goals made per game|
|`3p_made`|Average number of three-point field goals made per game|
|`3pa`|Average number of three-point field goal attempts per game|
|`3p`|Average percent of three-point field goals made per game|
|`ftm`|Average number of free throws made per game|
|`fta`|Average number of free throw attempts per game|
|`ft`|Average percent of free throws made per game|
|`oreb`|Average number of offensive rebounds per game|
|`dreb`|Average number of defensive rebounds per game|
|`reb`|Average number of rebounds per game|
|`ast`|Average number of assists per game|
|`stl`|Average number of steals per game|
|`blk`|Average number of blocks per game|
|`tov`|Average number of turnovers per game|
|`target_5yrs`|1 if career duration >= 5 yrs, 0 otherwise|

</center>

**Question:** Which columns would you select and avoid selecting as features, and why? Keep in mind the goal is to identify features that will serve as useful predictors for the target variable, `target_5yrs`. 

Columns to select:

gp (Number of games played): This is a relevant feature as players who have played more games are likely to have more experience and stability in their careers, which may be a predictor for career duration.

min (Number of minutes played): The amount of playing time is an important factor in career longevity, as more minutes may indicate a player's ability to contribute significantly to their team.

pts (Average number of points per game): Scoring ability is often linked to a player’s importance to the team, which could help predict career longevity.

fgm, fga, and fg (Field goals made, attempts, and percentage): These columns indicate shooting efficiency, which is a critical aspect of a player's performance and can be important for career longevity.

3p_made, 3pa, 3p (Three-point field goals made, attempts, and percentage): With the increasing importance of three-point shooting in modern basketball, these features could provide valuable insight.

ftm, fta, and ft (Free throws made, attempts, and percentage): Free throw ability is important and could serve as a predictor for player performance and career longevity.

reb, oreb, and dreb (Rebounds): Rebounding ability is a key performance indicator in basketball, which could relate to a player's long-term effectiveness.

ast (Average number of assists per game): Assists demonstrate a player’s ability to contribute to the team's offensive plays, which is important for team success and career longevity.

stl, blk (Steals and blocks): Defensive skills can also be predictive of a player’s effectiveness and career longevity.

Next, select the columns you want to proceed with. Make sure to include the target column, `target_5yrs`. Display the first few rows to confirm they are as expected.

In [10]:
# Select the columns to proceed with and save the DataFrame in a new variable 'selected_data'
selected_columns = ['gp', 'min', 'pts', 'fgm', 'fga', 'fg', '3p_made', '3pa', '3p', 'ftm', 'fta', 'ft', 'oreb', 'dreb', 'reb', 'ast', 'stl', 'blk', 'tov', 'target_5yrs']

selected_data = data[selected_columns]

# Display the first few rows
selected_data.head()


,gp,min,pts,fgm,fga,fg,3p_made,3pa,3p,ftm,fta,ft,oreb,dreb,reb,ast,stl,blk,tov,target_5yrs
0,36,27.4,7.4,2.6,7.6,34.7,0.5,2.1,25.0,1.6,2.3,69.9,0.7,3.4,4.1,1.9,0.4,0.4,1.3,0
1,35,26.9,7.2,2.0,6.7,29.6,0.7,2.8,23.5,2.6,3.4,76.5,0.5,2.0,2.4,3.7,1.1,0.5,1.6,0
2,74,15.3,5.2,2.0,4.7,42.2,0.4,1.7,24.4,0.9,1.3,67.0,0.5,1.7,2.2,1.0,0.5,0.3,1.0,0
3,58,11.6,5.7,2.3,5.5,42.6,0.1,0.5,22.6,0.9,1.3,68.9,1.0,0.9,1.9,0.8,0.6,0.1,1.0,1
4,48,11.5,4.5,1.6,3.0,52.4,0.0,0.1,0.0,1.3,1.9,67.4,1.0,1.5,2.5,0.3,0.3,0.4,0.8,1


<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the materials about feature selection and selecting a subset of a DataFrame.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Use two pairs of square brackets, and place the names of the columns you want to select inside the innermost brackets. 

</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

There is a function in `pandas` that can be used to display the first few rows of a DataFrame. Make sure to specify the column names with spelling that matches what's in the data. Use quotes to represent each column name as a string. 
</details>

### Feature transformation

An important aspect of feature transformation is feature encoding. If there are categorical columns that you would want to use as features, those columns should be transformed to be numerical. This technique is also known as feature encoding.

**Question:** Why is feature transformation important to consider? Are there any transformations necessary for the features you want to use?


**Feature transformation** is important because:

1. **Improving Model Performance**: Most machine learning algorithms require numerical data as input. Features like categorical variables (e.g., player position, team, etc.) must be transformed into numerical representations. Transformation helps the algorithm understand the relationships between features and the target variable.

2. **Handling Categorical Data**: Categorical features (like player position or team) need to be encoded into a numerical form (e.g., using techniques like one-hot encoding or label encoding) so that the machine learning algorithm can process them.

3. **Normalization/Standardization**: Some models (like linear regression or neural networks) may benefit from normalization or standardization of numerical features to scale them within a similar range. This ensures that one feature does not dominate the model simply because it has a larger numerical range.

4. **Dealing with Non-linearity**: Transforming certain features (e.g., using logarithmic transformations for highly skewed features) can make relationships more linear and improve model accuracy.

5. **Feature Engineering**: Sometimes creating new features (like combining two features, or performing arithmetic operations on them) can provide more predictive power than the individual features.

**Are transformations necessary for the features?**
- **Numerical Features**: For the selected numerical features (e.g., `gp`, `min`, `pts`, `fgm`, etc.), no transformations may be necessary unless there is a need for scaling (e.g., using normalization or standardization for some models like kNN or SVM).
- **Categorical Features**: If any categorical features (such as team or position) were present, they would need to be encoded numerically. However, since the dataset provided only contains numerical features (except for `name` which is not useful for modeling), no encoding is required for the selected features.
- **Scaling**: If using algorithms that are sensitive to feature scales (like k-NN or SVM), you may want to standardize the features (e.g., `gp`, `min`, `pts`, etc.) to a common scale, particularly if they have different units (minutes vs. points).


### Feature extraction

Display the first few rows containing containing descriptions of the data for reference. The table is as follows:

<center>

|Column Name|Column Description|
|:---|:-------|
|`name`|Name of NBA player|
|`gp`|Number of games played|
|`min`|Number of minutes played per game|
|`pts`|Average number of points per game|
|`fgm`|Average number of field goals made per game|
|`fga`|Average number of field goal attempts per game|
|`fg`|Average percent of field goals made per game|
|`3p_made`|Average number of three-point field goals made per game|
|`3pa`|Average number of three-point field goal attempts per game|
|`3p`|Average percent of three-point field goals made per game|
|`ftm`|Average number of free throws made per game|
|`fta`|Average number of free throw attempts per game|
|`ft`|Average percent of free throws made per game|
|`oreb`|Average number of offensive rebounds per game|
|`dreb`|Average number of defensive rebounds per game|
|`reb`|Average number of rebounds per game|
|`ast`|Average number of assists per game|
|`stl`|Average number of steals per game|
|`blk`|Average number of blocks per game|
|`tov`|Average number of turnovers per game|
|`target_5yrs`|1 if career duration >= 5 yrs, 0 otherwise|

</center>

In [11]:
# Select the columns to proceed with and save the DataFrame in a new variable 'selected_data'
selected_columns = ['gp', 'min', 'pts', 'fgm', 'fga', 'fg', '3p_made', '3pa', '3p', 'ftm', 'fta', 'ft', 'oreb', 'dreb', 'reb', 'ast', 'stl', 'blk', 'tov', 'target_5yrs']

selected_data = data[selected_columns]

# Display the first few rows
selected_data.head()




,gp,min,pts,fgm,fga,fg,3p_made,3pa,3p,ftm,fta,ft,oreb,dreb,reb,ast,stl,blk,tov,target_5yrs
0,36,27.4,7.4,2.6,7.6,34.7,0.5,2.1,25.0,1.6,2.3,69.9,0.7,3.4,4.1,1.9,0.4,0.4,1.3,0
1,35,26.9,7.2,2.0,6.7,29.6,0.7,2.8,23.5,2.6,3.4,76.5,0.5,2.0,2.4,3.7,1.1,0.5,1.6,0
2,74,15.3,5.2,2.0,4.7,42.2,0.4,1.7,24.4,0.9,1.3,67.0,0.5,1.7,2.2,1.0,0.5,0.3,1.0,0
3,58,11.6,5.7,2.3,5.5,42.6,0.1,0.5,22.6,0.9,1.3,68.9,1.0,0.9,1.9,0.8,0.6,0.1,1.0,1
4,48,11.5,4.5,1.6,3.0,52.4,0.0,0.1,0.0,1.3,1.9,67.4,1.0,1.5,2.5,0.3,0.3,0.4,0.8,1


**Question:** Which columns lend themselves to feature extraction?

Feature extraction is the process of creating new features from existing ones to better capture the underlying patterns in the data. In this case, the following columns lend themselves to feature extraction:

1. **`fgm`, `fga`, and `fg`**: 
   - These columns provide information about field goal performance (made, attempted, and percentage). You can create new features such as:
     - **Field Goal Efficiency**: A ratio of `fgm` to `fga` or `fg` to `100` (field goal percentage).
     - **Field Goal Attempts per Game**: A new feature based on `fga` that might be relevant for understanding shooting volume and efficiency.

2. **`3p_made`, `3pa`, and `3p`**:
   - These columns provide insights into three-point shooting. Possible new features could be:
     - **3-point shooting efficiency**: A ratio of `3p_made` to `3pa` or `3p` to `100`.
     - **Three-Point Usage**: A feature that indicates the player's tendency to attempt three-point shots, potentially derived from the ratio of `3pa` to `fga`.

3. **`ftm`, `fta`, and `ft`**:
   - These columns give information about free throw performance. New features could be:
     - **Free Throw Efficiency**: A ratio of `ftm` to `fta` or `ft` to `100`.
     - **Free Throw Attempts per Game**: A new feature derived from `fta`.

4. **`oreb`, `dreb`, and `reb`**:
   - These columns provide data on different types of rebounds. You could create new features like:
     - **Rebound Rate**: A ratio of total rebounds (`reb`) to minutes played (`min`).
     - **Offensive vs. Defensive Rebound Split**: A new feature representing the ratio of `oreb` to `dreb`.

5. **`ast`**:
   - Assists are an important factor in basketball, and you could derive features like:
     - **Assist-to-Turnover Ratio**: A ratio of `ast` to `tov`, which is a key metric for evaluating playmaking efficiency.

6. **`stl`, `blk`**:
   - Steals and blocks are crucial defensive metrics. A new feature could be:
     - **Defensive Efficiency**: Combining `stl` and `blk` to measure a player's overall defensive contribution.

7. **`tov`**:
   - Turnovers are important in evaluating a player's decision-making. You could combine it with assists to create an **assist-to-turnover ratio** or track it as a feature by itself to measure ball control.

By creating new features from these existing ones, you could better capture the relationships between various performance metrics, which may help improve predictive modeling for the target variable (`target_5yrs`).

Let me know if you'd like assistance with the code to create these features!

Extract two features that you think would help predict `target_5yrs`. Then, create a new variable named 'extracted_data' that contains features from 'selected_data', as well as the features being extracted.

In [12]:
# Select the columns to proceed with and save the DataFrame in a new variable 'selected_data'
selected_columns = ['gp', 'min', 'pts', 'fgm', 'fga', 'fg', '3p_made', '3pa', '3p', 'ftm', 'fta', 'ft', 'oreb', 'dreb', 'reb', 'ast', 'stl', 'blk', 'tov', 'target_5yrs']
selected_data = data[selected_columns]

# Extract two features that could help predict target_5yrs
# Feature 1: Field Goal Efficiency (fgm/fga)
selected_data['fg_efficiency'] = selected_data['fgm'] / selected_data['fga']

# Feature 2: Assist-to-Turnover Ratio (ast/tov)
selected_data['ast_to_tov'] = selected_data['ast'] / selected_data['tov']

# Create a new variable 'extracted_data' containing the original features as well as the extracted features
extracted_data = selected_data.copy()

# Display the first few rows to confirm
extracted_data.head()


,gp,min,pts,fgm,fga,fg,3p_made,3pa,3p,ftm,...,oreb,dreb,reb,ast,stl,blk,tov,target_5yrs,fg_efficiency,ast_to_tov
0,36,27.4,7.4,2.6,7.6,34.7,0.5,2.1,25.0,1.6,...,0.7,3.4,4.1,1.9,0.4,0.4,1.3,0,0.342105,1.461538
1,35,26.9,7.2,2.0,6.7,29.6,0.7,2.8,23.5,2.6,...,0.5,2.0,2.4,3.7,1.1,0.5,1.6,0,0.298507,2.312500
2,74,15.3,5.2,2.0,4.7,42.2,0.4,1.7,24.4,0.9,...,0.5,1.7,2.2,1.0,0.5,0.3,1.0,0,0.425532,1.000000
3,58,11.6,5.7,2.3,5.5,42.6,0.1,0.5,22.6,0.9,...,1.0,0.9,1.9,0.8,0.6,0.1,1.0,1,0.418182,0.800000
4,48,11.5,4.5,1.6,3.0,52.4,0.0,0.1,0.0,1.3,...,1.0,1.5,2.5,0.3,0.3,0.4,0.8,1,0.533333,0.375000


<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the materials about feature extraction.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Use the function `copy()` to make a copy of a DataFrame. To access a specific column from a DataFrame, use a pair of square brackets and place the name of the column as a string inside the brackets.

</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Use a pair of square brackets to create a new column in a DataFrame. The columns in DataFrames are series objects, which support elementwise operations such as multiplication and division. Be sure the column names referenced in your code match the spelling of what's in the DataFrame.
</details>

Now, to prepare for the Naive Bayes model that you will build in a later lab, clean the extracted data and ensure ensure it is concise. Naive Bayes involves an assumption that features are independent of each other given the class. In order to satisfy that criteria, if certain features are aggregated to yield new features, it may be necessary to remove those original features. Therefore, drop the columns that were used to extract new features.

**Note:** There are other types of models that do not involve independence assumptions, so this would not be required in those instances. In fact, keeping the original features may be beneficial.

In [13]:
# Select the columns to proceed with and save the DataFrame in a new variable 'selected_data'
selected_columns = ['gp', 'min', 'pts', 'fgm', 'fga', 'fg', '3p_made', '3pa', '3p', 'ftm', 'fta', 'ft', 'oreb', 'dreb', 'reb', 'ast', 'stl', 'blk', 'tov', 'target_5yrs']
selected_data = data[selected_columns]

# Extract two features that could help predict target_5yrs
# Feature 1: Field Goal Efficiency (fgm/fga)
selected_data['fg_efficiency'] = selected_data['fgm'] / selected_data['fga']

# Feature 2: Assist-to-Turnover Ratio (ast/tov)
selected_data['ast_to_tov'] = selected_data['ast'] / selected_data['tov']

# Create a new variable 'extracted_data' containing the original features as well as the extracted features
extracted_data = selected_data.copy()

# Now drop the columns used to create new features ('fgm', 'fga', 'ast', 'tov')
extracted_data = extracted_data.drop(columns=['fgm', 'fga', 'ast', 'tov'])

# Display the first few rows of 'extracted_data' to ensure that column drops took place
extracted_data.head()


,gp,min,pts,fg,3p_made,3pa,3p,ftm,fta,ft,oreb,dreb,reb,stl,blk,target_5yrs,fg_efficiency,ast_to_tov
0,36,27.4,7.4,34.7,0.5,2.1,25.0,1.6,2.3,69.9,0.7,3.4,4.1,0.4,0.4,0,0.342105,1.461538
1,35,26.9,7.2,29.6,0.7,2.8,23.5,2.6,3.4,76.5,0.5,2.0,2.4,1.1,0.5,0,0.298507,2.312500
2,74,15.3,5.2,42.2,0.4,1.7,24.4,0.9,1.3,67.0,0.5,1.7,2.2,0.5,0.3,0,0.425532,1.000000
3,58,11.6,5.7,42.6,0.1,0.5,22.6,0.9,1.3,68.9,1.0,0.9,1.9,0.6,0.1,1,0.418182,0.800000
4,48,11.5,4.5,52.4,0.0,0.1,0.0,1.3,1.9,67.4,1.0,1.5,2.5,0.3,0.4,1,0.533333,0.375000


<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the materials about feature extraction.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

There are functions in the `pandas` library that remove specific columns from a DataFrame and that display the first few rows of a DataFrame.
</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Use the `drop()` function and pass in a list of the names of the columns you want to remove. By default, calling this function will result in a new DataFrame that reflects the changes you made. The original DataFrame is not automatically altered. You can reassign `extracted_data` to the result, in order to update it. 

Use the `head()` function to display the first few rows of a DataFrame.
</details>

Next, export the extracted data as a new .csv file. You will use this in a later lab. 

In [14]:
# Select the columns to proceed with and save the DataFrame in a new variable 'selected_data'
selected_columns = ['gp', 'min', 'pts', 'fgm', 'fga', 'fg', '3p_made', '3pa', '3p', 'ftm', 'fta', 'ft', 'oreb', 'dreb', 'reb', 'ast', 'stl', 'blk', 'tov', 'target_5yrs']
selected_data = data[selected_columns]

# Extract two features that could help predict target_5yrs
# Feature 1: Field Goal Efficiency (fgm/fga)
selected_data['fg_efficiency'] = selected_data['fgm'] / selected_data['fga']

# Feature 2: Assist-to-Turnover Ratio (ast/tov)
selected_data['ast_to_tov'] = selected_data['ast'] / selected_data['tov']

# Create a new variable 'extracted_data' containing the original features as well as the extracted features
extracted_data = selected_data.copy()

# Remove the columns that were used to extract new features (e.g., 'fgm', 'fga', 'ast', 'tov')
extracted_data = extracted_data.drop(columns=['fgm', 'fga', 'ast', 'tov'])

# Export the extracted data to a new .csv file
extracted_data.to_csv('extracted_nba_players_data.csv', index=False)



<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

There is a function in the `pandas` library that exports a DataFrame as a .csv file. 
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Use the `to_csv()` function to export the DataFrame as a .csv file. 
</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Call the `to_csv()` function on `extracted_data`, and pass in the name that you want to give to the resulting .csv file. Specify the file name as a string and in the file name. Make sure to include `.csv` as the file extension. Also, pass in the parameter `index` set to `0`, so that when the export occurs, the row indices from the DataFrame are not treated as an additional column in the resulting file. 
</details>

## **Considerations**


**What are some key takeaways that you learned during this lab? Consider the process you followed and what tasks were performed during each step, as well as important priorities when training data.**

### Key Takeaways from This Lab:

1. **Understanding Data Structure**:
   - **Exploratory Data Analysis (EDA)**: It’s crucial to start by getting a sense of the data’s structure, such as the number of rows, columns, and column names. This allows us to understand what we’re working with and ensures that we use the appropriate columns for analysis or modeling.
   - **Identifying Target Variable**: Understanding the target variable (`target_5yrs` in this case) and its relevance is essential. Knowing whether the dataset is balanced or imbalanced helps in selecting the right approach and model for prediction.

2. **Feature Engineering**:
   - **Feature Extraction**: This involves creating new features from the existing ones, which can help improve model performance. For example, we created **Field Goal Efficiency** and **Assist-to-Turnover Ratio**. These transformations can provide better insights and increase the predictive power of the model.
   - **Feature Transformation**: If any features are categorical, transforming them into numerical formats (like encoding) is a necessary step for machine learning models that require numerical input.
   - **Feature Removal**: In the case of Naive Bayes (which assumes feature independence), it’s important to remove features that were used to create new ones. This ensures the model can satisfy the assumption of feature independence and avoids potential data leakage.

3. **Data Cleaning**:
   - **Handling Missing Values**: Checking for missing values is a crucial part of data preprocessing. Missing data can affect model performance if not handled appropriately. Depending on the amount and type of missing data, we may decide to drop rows/columns or impute values.
   - **Data Export**: Once the necessary transformations and cleaning are done, it's important to save and export the data in a suitable format (like `.csv`) for future use. This step ensures that the cleaned data is available for further analysis or model training.

4. **Data Preprocessing for Modeling**:
   - **Scaling and Normalization**: While this was not covered directly in the lab, it’s important to mention that certain algorithms (like k-NN or SVM) might require feature scaling or normalization to prevent features with larger scales from dominating the model.
   - **Feature Selection vs. Feature Engineering**: Selecting the right features for the model is critical. In the case of Naive Bayes, we removed the features that were used to create new ones, ensuring that our features are independent and meet the model's assumptions.

5. **Key Priorities in Training Data**:
   - **Quality of Data**: The quality of the training data is one of the most important factors influencing model performance. Ensuring that the data is clean, complete, and properly transformed is crucial for building an effective model.
   - **Relevance of Features**: Choosing features that have a meaningful relationship with the target variable is important. Irrelevant or redundant features can reduce the model's performance and increase the complexity.
   - **Class Balance**: Checking the class balance is critical, especially when working with classification problems. If the target variable is imbalanced, resampling techniques or careful evaluation using metrics like recall and precision may be necessary.

### Final Thoughts:
This lab reinforced the importance of **data cleaning**, **feature engineering**, and **preprocessing** in preparing data for machine learning. By transforming, selecting, and preparing the right features, we increase the likelihood of building more accurate and efficient predictive models.

**What summary would you provide to stakeholders? Consider key attributes to be shared from the data, as well as upcoming project plans.**

**Overview of the Data**:
We have gathered a subset of NBA player data that includes both performance metrics and information on whether a player’s career lasted for at least five years. The dataset contains **1,340 entries** and **21 features**, with key performance indicators such as **points per game (pts)**, **assists (ast)**, **field goals made (fgm)**, and **rebounds (reb)**. This data will help us understand which factors most strongly correlate with career longevity in the NBA.

**Key Insights from the Data**:
- **Numerical Features**: The data contains a wide array of numerical performance metrics that describe the efficiency and contribution of NBA players, including shooting percentages (field goals, three-point shots, free throws), rebounds, assists, steals, and blocks.
- **Target Variable**: The target variable (`target_5yrs`) indicates whether a player's career lasted 5 years or more, which will be used to predict the likelihood of career longevity.
- **Feature Engineering**: We created new features such as **Field Goal Efficiency** and **Assist-to-Turnover Ratio** to capture more relevant aspects of a player’s performance. These features will help improve the predictive power of the model.
- **Data Cleanliness**: The data was reviewed for missing values, and we found that all columns contained complete data, making the dataset ready for analysis.

**Next Steps**:
- **Feature Selection**: We will continue refining our features and removing any redundant ones to ensure we’re working with the most relevant predictors.
- **Modeling**: The next stage involves using this data to train a **Naive Bayes model** to predict whether a player’s career will last at least five years, based on their performance metrics. We will need to check if the features satisfy the independence assumptions of Naive Bayes or consider other algorithms if necessary.
- **Model Evaluation**: After training the model, we will evaluate its performance using accuracy, precision, recall, and F1-score to ensure reliable predictions.
- **Actionable Insights**: The insights from this model can be used to identify players most likely to succeed in the long term, allowing teams to make more informed decisions when scouting and recruiting players.

**Impact for Stakeholders**:
This analysis can provide valuable data-driven insights for NBA managers and coaches, helping them identify promising players who are likely to thrive in the league over time. These insights could inform decisions related to player acquisitions, training programs, and long-term team building strategies.

We are excited about the upcoming steps, and with further modeling, we aim to deliver actionable predictions that could influence key decision-making in player recruitment and team development.

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.